pkg/apis/apps/types.go

```diff
@@ -415,6 +415,28 @@ type DeploymentSpec struct {
 	// the max value of int32 (i.e. 2147483647) by default, which means "no deadline".
 	// +optional
 	ProgressDeadlineSeconds *int32
+
+	// podReplacementPolicy specifies when to create replacement Pods.
+	// Possible values are:
+	// - TerminationStarted policy creates replacement Pods when the old Pods start
+	//   terminating (has a .metadata.deletionTimestamp). The total number of
+	//   Deployment Pods can be greater than specified by the Deployment's
+	//   .spec.replicas and the DeploymentStrategy.
+	// - TerminationComplete policy creates replacement Pods only when the old Pods
+	//   are fully terminated (reach Succeeded or Failed phase). The old Pods are
+	//   subsequently removed. The total number of the Deployment Pods is
+	//   limited by the Deployment's .spec.replicas and the DeploymentStrategy.
+	//
+	// The default behavior when the policy is not specified depends on the DeploymentStrategy:
+	// - Recreate strategy uses TerminationComplete behavior when recreating the deployment,
+	//   but uses TerminationStarted when scaling the deployment.
+	// - RollingUpdate strategy uses TerminationStarted behavior for both rolling out and
+	//   scaling the deployments.
```

Comment:
>is the scaling aspect only relevant when scaling happens in the middle of a rollout?

No it also checks for terminating pods when scaling (scale up) during a normal operation.

For example if you scale down 5 -> 2 and then scale up again 2 -> 5, it will wait for the pods to terminate (which can result in a substantial delay).  If pods terminate also for other reasons, it will wait with scaling up, until there is a capacity to run those pods.

> What does a replicaset in steady state do when one of its pods starts terminating? doesn't it replace immediately? If so, this field does not actually guarantee that the number of pods for the deployment will not exceed spec.replicas if maxSurge is 0, right?

Yes, this is only for situations where Deployment is managed, e.g. during rollouts or scaling. If there is an external disruption, it can still exceed the number of desired pods.

I mention this limitation in the docs PR: https://github.com/kubernetes/website/pull/45266. I suppose I should mention something similar here as well, right?

Reply:
> Yes, this is only for situations where the Deployment is managed, e.g. during rollouts or scaling. If there is an external disruption, it may still exceed the number of desired pods.

Isn't that just as much of a problem for triggering autoscaling in undesirable ways? Building up from primitives, if replicasets could be told not to exceed their number of replicas for non-terminated pods when replacing pods, how would that affect the deployment portion of this enhancement, which builds on top of capabilities offered by replicasets?

Reply:
Unfortunately, implementing this feature for ReplicaSets might have undesirable/unexpected consequences for the user. One of these is the minimum availability guaranteed by the Deployment (`.spec.replicas - .spec.strategy.rollingUpdate.maxUnavailable`. Even though we document `maxUnavailable` as a field for updates, I think people expect this to hold true throughout the whole lifecycle of the deployment. I've heard similar misunderstandings before. 

So for example, if you have a ReplicaSet, say, with 1-3 pods running on the same node, and something disrupts/evicts all of the pods. It could actually take a long time for even a single replica to become available again, resulting in application downtime.


I have included a general rationale in the KEP as to why I think it is not a good idea to implement this for RSs: https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3973-consider-terminating-pods-deployment/README.md#consideration-for-other-controllers

Regarding the balancing, the RS can have extra pods during the down scale of the RS, so the deployment controller has to make the number of pods compliant across all RSs. So the invariant can still be broken for a single RS. A RS is a simple primitive and has no concept of a rollout.

I think the main issue bothering people today are the extra pods in rollouts, which this PR would help with. Unfortunately, I do not see an elegant solution for the external disruptions.